### PR TITLE
Integrate some modules to enhance tiny defects

### PR DESCRIPTION
**Pull request by GiaThinh110605**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds Inner-CIoU bounding-box loss support alongside experimental detection heads and attention modules.

### 📊 Key Changes
- Replaces the standard CIoU calculation in `BboxLoss` with a new `inner_ciou` implementation.
- Adds tests covering Inner-CIoU loss values, gradients, and bbox-loss execution.
- Introduces AHFIN, MECS, MultiScaleGhost, AnchorFreeHead, and HybridHead modules with parser integration.
- Adds model parsing and hybrid-head handling in `tasks.py`, including custom stride and loss extraction logic.
- Adds an experimental Wasserstein distance utility and an empty `test__` file.

### 🎯 Purpose & Impact
- Enables experimentation with Inner-IoU-based box regression and additional detection architectures.
- Expands model configuration support for multi-input feature fusion and hybrid detection heads.
- May affect training behavior and loss convergence for all detection models because the default bbox loss is changed globally.
- Requires further validation for device portability, model integration, formatting, and compatibility with existing training and inference paths.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @GiaThinh110605, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. This automated message will be followed up by an engineer who can assist with the review. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/AHFIN.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    def forward(self, x_list):
+        # 对每个特征图进行卷积和尺度调整
+        scaled_x_list = []
+        for i in range(self.num_inputs):
+            scaled_x = self.convs[i](x_list[i]) * self.scales[i]
+            scaled_x_list.append(scaled_x)
+
+        # 特征融合
+        fused_x = sum(scaled_x_list)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/AHFIN.py:38]: ⚠️ **HIGH**: This fusion assumes every input already has identical spatial dimensions, but the added parser test wires outputs from stride-2 and stride-4 convolutions into this module. Running that parsed model reaches this line with 128x128 and 64x64 tensors and raises a size-mismatch error. Resize/project inputs to a common resolution, or reject incompatible inputs during model parsing and cover an actual forward pass in the test.

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -591,6 +626,21 @@ def _clip_augmented(self, y):
...
+        HybridHead returns {"anchor": {"boxes":..., "scores":..., "feats":[...]}, "anchor_free": [...]}
+        but v8DetectionLoss expects the anchor dict directly. Extract it here.
+        """
+        if getattr(self, "criterion", None) is None:
+            self.criterion = self.init_criterion()
+        if preds is None:
+            preds = self.forward(batch["img"])
+        # HybridHead wraps anchor and anchor_free in a top-level dict
+        if isinstance(preds, dict) and "anchor" in preds and "anchor_free" in preds:
+            preds = preds["anchor"]
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:641]: ⚠️ **HIGH**: This discards every `anchor_free` prediction before loss computation. Consequently the anchor-free branch receives no gradients during training; inference also returns only the anchor branch. The HybridHead therefore adds substantial forward/backward memory and compute without affecting the trained or deployed model. Add a compatible loss/inference merge for this branch, or do not construct/run it.

File: ultralytics/nn/tasks.py [hunk 3]
```diff
@@ -2009,6 +2066,8 @@ def parse_model(d, ch, verbose=True):
             args.append([ch[x] for x in f])  # nc, ch tuple
         elif m is v10Detect:
             args.append([ch[x] for x in f])
+        elif m in frozenset({HybridHead, AnchorFreeHead}):
+            args.append([ch[x] for x in f])
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:2070]: ⚠️ **HIGH**: `AnchorFreeHead` has the signature `(ch, nc)`, but this branch appends a channel list to the YAML arguments. With the normal detection-head YAML form `[nc]`, this instantiates `AnchorFreeHead(nc, [channels...])`, so `nn.Conv2d(..., out_channels=nc)` receives a list and model parsing fails. Handle `AnchorFreeHead` separately and construct its arguments in `(input_channels, nc)` order.

File: ultralytics/nn/modules/hybridhead.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license
+
+import torch.nn as nn
+
+from .anchor_free import AnchorFreeHead
+from .head import Detect
+
+
+class HybridHead(nn.Module):
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/hybridhead.py:9]: 💡 **MEDIUM**: This wrapper does not expose `nl`, although `DetectionModel._clip_augmented()` reads `self.model[-1].nl`. Since `HybridHead` is accepted as a detection head and does not set `end2end`, calling prediction with `augment=True` reaches that code path and raises `AttributeError`. Delegate `nl` to `anchor_head` along with the other required Detect attributes.

File: ultralytics/utils/inner_iou.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license
+
+import math
+
+import torch
+
+# code reference:https://github.com/malagoutou/Inner-IoU/blob/main/innerciou.py
+
+
+def inner_ciou(box1, box2, ratio=1.0, xywh=True, eps=1e-7):
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/inner_iou.py:10]: 💡 **MEDIUM**: With `ratio=1.0`, the inner boxes are identical to the original boxes, making `inner_iou` equal ordinary IoU and the returned value equivalent to the existing CIoU calculation. `BboxLoss` always uses this default and exposes no way to choose another ratio, so the global replacement does not actually enable an Inner-IoU loss.

File: ultralytics/utils/inner_iou.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    print("using inner-ciou")
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/inner_iou.py:11]: 💡 **MEDIUM**: `inner_ciou()` is called once per bbox-loss forward pass, so this emits a stdout line for every training batch. Besides flooding logs, synchronous console I/O can materially slow training. Remove this debug print.

File: tests/test_bbox_loss.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            target_scores=target_scores,
+            target_scores_sum=target_scores_sum,
+            fg_mask=fg_mask,
+            imgsz=imgsz,
+            stride=stride,
+        )
+        print("Success!")
+        print(f"loss_iou: {loss_iou.item()} (shape: {loss_iou.shape})")
+        print(f"loss_dfl: {loss_dfl.item()} (shape: {loss_dfl.shape})")
+    except Exception:
```
Comment (UltralyticsAssistant) [inline tests/test_bbox_loss.py:62]: 📝 **LOW**: The exception is printed and then swallowed, so this test passes even if `BboxLoss.forward()` raises. Remove the catch or re-raise after logging so the added regression test can fail CI when the loss path breaks.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR has several blocking integration issues in the experimental heads: AHFIN cannot run with the supplied multi-scale inputs, AnchorFreeHead is parsed with incompatible arguments, and HybridHead's anchor-free branch is never used for loss or inference. The new loss path also needs debug output removed and a configurable non-default Inner-IoU ratio.

💬 Posted 7 inline comments

**Comment by glenn-jocher** [comment]
Thank you for the contribution. I am closing this version because it combines unrelated experimental heads, attention blocks, Inner-CIoU, and Wasserstein utilities across 13 files, while globally replacing the default bbox loss. It is also not functional as submitted: AHFIN receives incompatible feature sizes, AnchorFreeHead parser arguments are reversed, the HybridHead anchor-free output is computed but discarded from both loss and inference, augmented inference lacks required head metadata, and one regression test swallows failures. This cannot be repaired without redesigning and splitting the proposal. Please submit one independently benchmarked module or loss per PR, with its complete train/inference integration, focused tests, and a literal `Deleted:` line. Thank you for the work and for understanding the scope bar.
